### PR TITLE
DAOS-4021 build: fix python for libduns vs lustre build on debian

### PR DESCRIPTION
**Pull request by bfaccini**
SConscript/python code to handle Lustre case was buggy.

Change-Id: If0e0dbc5dd5fd9ce0098761375fd9bbfe7b4738e
Signed-off-by: Bruno Faccini <bruno.faccini@intel.com>

### CODE REVIEW

File: src/client/dfs/SConscript [hunk 1]
```diff
@@ -7,49 +7,66 @@ def scons():
...
+            if gotpkg == True:
+                if parse_version(instver.version) >= parse_version('2.12.57'):
+                    gotversion = True
+        if gotdistro == False:
             print("Unable to identify distro.")
+        else:
+            if gotpkg == False:
+                print("Unable to find lustre nor lustre-client package.")
+            else:
+                if gotversion == False:
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: done

File: src/client/dfs/SConscript [hunk 2]
```diff
@@ -7,49 +7,66 @@ def scons():
...
+                    print("Unable to find minimum version of lustre nor lustre-client package.")
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: keep msg strings on a single line to allow them to be retrieved when browsing sources

File: src/client/dfs/SConscript [hunk 3]
```diff
@@ -7,49 +7,66 @@ def scons():
...
+    
+            gotdistro = True
             ts = rpm.TransactionSet()
             headers = ts.dbMatch('name', 'lustre')
             if not headers:
                 headers = ts.dbMatch('name', 'lustre-client')
             if headers:
                 from pkg_resources import parse_version
-
+
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: done

File: src/client/dfs/SConscript [hunk 4]
```diff
@@ -7,49 +7,66 @@ def scons():
...
-                        denv.AppendUnique(CCFLAGS=['-DLUSTRE_INCLUDE'])
-                        gotit = True
-            if gotit == False:
-                print("Found no compatible lustre nor lustre-client package.")
+                        gotversion = True
         elif distro in ('debian', 'ubuntu'):
             import apt
             from pkg_resources import parse_version
-
+
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: done

File: src/client/dfs/SConscript [hunk 5]
```diff
@@ -7,49 +7,66 @@ def scons():
...
+                    if cache["lustre-client"].is_installed:
+                        gotpkg = True
+                        instver = cache["lustre-client"].installed
+            if gotpkg == True:
+                if parse_version(instver.version) >= parse_version('2.12.57'):
+                    gotversion = True
+        if gotdistro == False:
             print("Unable to identify distro.")
+        else:
+            if gotpkg == False:
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: done

File: src/client/dfs/SConscript [hunk 6]
```diff
@@ -7,49 +7,66 @@ def scons():
...
-                print("Found no compatible lustre nor lustre-client package.")
-        else:
+                if "lustre-client" in cache:
+                    if cache["lustre-client"].is_installed:
+                        gotpkg = True
+                        instver = cache["lustre-client"].installed
+            if gotpkg == True:
+                if parse_version(instver.version) >= parse_version('2.12.57'):
+                    gotversion = True
+        if gotdistro == False:
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: done

File: src/client/dfs/SConscript [hunk 7]
```diff
@@ -7,49 +7,66 @@ def scons():
...
+                    gotpkg = True
+                    instver = cache["lustre"].installed
             else:
-                print("Found no compatible lustre nor lustre-client package.")
-        else:
+                if "lustre-client" in cache:
+                    if cache["lustre-client"].is_installed:
+                        gotpkg = True
+                        instver = cache["lustre-client"].installed
+            if gotpkg == True:
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: done

File: src/client/dfs/SConscript [hunk 8]
```diff
@@ -7,49 +7,66 @@ def scons():
...
+                if re.search('^ID=', line):
                     line = line.strip()
                     _, distro = line.strip().split('=')
                     distro = distro.strip('"')
                     break
         if distro in ['centos', 'rhel', 'fedora', 'opensuse']:
             import rpm
-
-            gotit = False
+
```
[the remark this answers is not in the document]
Comment (bfaccini) [reply src/client/dfs/SConscript]: line no longer present

File: src/client/dfs/SConscript [hunk 9]
```diff
@@ -1,5 +1,7 @@
 """Build FUSE client"""
 import daos_build
+import os.path
+from os import path
```
Comment (brianjmurrell) [inline src/client/dfs/SConscript]: Isn't this effectively double-importing `os.path` into two different namespaces?

File: src/client/dfs/SConscript [hunk 10]
```diff
@@ -54,19 +56,45 @@ def scons():
                         instver = cache["lustre-client"].installed
             if gotpkg is True:
                 if parse_version(instver.version) >= parse_version('2.12.57'):
```
Comment (brianjmurrell) [inline src/client/dfs/SConscript]: This would be more robustly done as feature tests rather than a version comparison.  Version numbers can be misleading.  The EL7 kernel is an excellent example of this.  It's version number is a **very** unreliable way of knowing that kernel supports.

File: src/client/dfs/SConscript [hunk 11]
```diff
@@ -54,19 +56,45 @@ def scons():
...
                 else:
-                    denv.AppendUnique(CCFLAGS=['-DLUSTRE_INCLUDE'])
-    except ImportError:
-        raise
+                    print("Installed Lustre version is compatible.")
+    except ImportError as error:
+        print(error.message +
+              ", trying to find Lustre headers in usual locations.")
+        """ Try the empirical way, by just trying to check both required header files are present ... """
+        if path.exists("/usr/include/lustre/lustreapi.h"):
```
Comment (brianjmurrell) [inline src/client/dfs/SConscript]: This is the **only** kind of test that I think this code should be doing IMHO.  It doesn't matter how (RPM, deb, `make install`, etc.) the header files are present, just that they are.  There should be an option to scons to also tell it where the headers are in case they are in a non-standard location.

But frankly, the presence of header files is not usually enough.  You need to know that they represent a compatible API.  For that you typically do API feature compile tests and `scons` provides tools to do that kind of checking without having to write our own python code to do it.

This code should be using those features and only those features for determining the _lustre-client_ compatibility.  Doing that would make this [whole section of code](https://github.com/daos-stack/daos/blob/a3d74025fc61315b1aeffbeb790c98cbd6fcc2a4/src/client/dfs/SConscript#L12-L105) much easier to read, understand and maintain.

File: src/client/dfs/SConscript [hunk 12]
```diff
@@ -54,19 +56,45 @@ def scons():
...
+            if path.exists("/usr/include/linux/lustre/lustre_user.h"):
+                gotpkg = True
+                with open('/usr/include/linux/lustre/lustre_user.h', 'r') as file:
+                    for line in file:
+                        """ if definition present Lustre version is compatible """
+                        if 'LU_FOREIGN_TYPE_SYMLINK' in line:
```
Comment (brianjmurrell) [inline src/client/dfs/SConscript]: This is a typical example of an API feature test that you can have `scons` do.  This code should be using that, not writing it's own python to guesstimate.

File: src/client/dfs/SConscript (hunk 9 above)
Comment (bfaccini) [reply src/client/dfs/SConscript]: right!

File: src/client/dfs/SConscript (hunk 10 above)
Comment (bfaccini) [reply src/client/dfs/SConscript]: > This would be more robustly done as feature tests

Not sure to understand this. Do you mean it should be tested by trying to compile a small snippet including the necessary Lustre headers and using/calling the specific defines/APIs from the Lustre feature we are looking for ?? Like frequently done with configure/gnu-based builds ??

File: src/client/dfs/SConscript (hunk 11 above)
Comment (bfaccini) [reply src/client/dfs/SConscript]: > scons provides tools to do that kind of checking without having to write our own python code to do it.

Can you be more detailed, or may be point to some existing in our build ?

File: src/client/dfs/SConscript (hunk 12 above)
Comment (bfaccini) [reply src/client/dfs/SConscript]: Well, this is where my "scons" knowledge finds it is very low limit... BTW, do you know what a scons is in french, but may be you also use this name?

File: src/client/dfs/SConscript (hunk 11 above)
Comment (brianjmurrell) [reply src/client/dfs/SConscript]: There is a whole [manual for scons](https://scons.org/doc/0.97/HTML/scons-user/book1.html) that details all of the things it can do like [header checks](https://scons.org/doc/0.97/HTML/scons-user/x2827.html), [function availability](https://scons.org/doc/0.97/HTML/scons-user/x2836.html), and the ability to [add custom checks](https://scons.org/doc/0.97/HTML/scons-user/x2867.html).  Among many other things.

These are the tools of scons for determining the capabilities of APIs, without having to write a whole bunch of python code that will need ongoing maintenance.

File: src/client/dfs/SConscript (hunk 12 above)
Comment (brianjmurrell) [reply src/client/dfs/SConscript]: But Scons is the tool of choice for this project.  For maintenance (in the least) we should be using the features of of the tool of choice and not end up rewriting it in free-flowing Python code that will increase the maintenance burden.

File: src/client/dfs/SConscript (hunk 10 above)
Comment (brianjmurrell) [reply src/client/dfs/SConscript]: Yes, that is exactly what I mean.  But I don't mean you should write free-form python code to do it but rather use the features built into SCons to do it for you.  SCons is analogous to "configure/gnu-based" (actually called GNU Autoconf or more generally GNU Autotools) builds and provides the same kind of tools for API discovery as Autoconf.

See my next comment for more details.

**Comment by bfaccini** [comment]
> Isn't lines 16-25 re-inventing [python-distro](https://pypi.org/project/distro/) with the added burden of having to maintain the re-invention as distros evolve?

Right. This is the strength of python, there is always an existing module already doing what you want, but for me this is also a weakness because sometimes you may have to be patient and spend a lot of time googling for it and then also to understand its behaviour, and I am not patient enough...

**Comment by brianjmurrell** [comment]
> Right. This is the strength of python, there is always an existing module already doing what you want, but for me this is also a weakness because sometimes you may have to be patient and spend a lot of time googling for it and then also to understand its behaviour, and I am not patient enough...

Well, it only took about 30 seconds (or less) of Googling to find that.  But now that you know about it, wouldn't it be better to utilise maintained code for that functionality rather than re-writing it and having to maintain it as distros mature and change?

**Comment by jolivier23** [comment]
much better!

**Comment by bfaccini** [comment]
> much better!

@jolivier23 I am trying to improve my python/scons skills, with the highly valuable assistance and great patience of @brianjmurrell !! ;-)

**Comment by bfaccini** [comment]
hmmm, looks like the CI failures reported are known issues already reported as DAOS-3945 and DAOS-4294/DAOS-4285 ....

**Comment by bfaccini** [comment]
> With this change, should we be actually building and testing the lustre-daos combination?
> How do we know this works and doesn't bitrot if we don't?

Well you are right this is a bit tricky to test all cases for this change, I already did the testing on a local build+install with/without Lustre version and compatible or not installed. But re-doing so I have just found a problem (original environment is "replicated" in configure context so to modify original Configure.env reference must be used) that will be fixed in next push.
For a RPM-based install and using the artifacts from CI build I can only test the case where no Lustre or compatible install is detected (as you have no or incompatible Lustre on builders)... Is it possible to generate RPMs in a local build ? And if yes, howto?

bitrot, hmm what do you mean?? In french I know bistrot (a bar in slang !!) ...

BTW, does the new style of my changes, more python/scons oriented, already looks better for you ??

**Comment by brianjmurrell** [comment]
> Well you are right this is a bit tricky to test all cases for this change,

But surely we could be testing the main case which is to build lustre-enabled, couldn't we?

> For a RPM-based install and using the artifacts from CI build I can only test the case where no Lustre or compatible install is detected (as you have no or incompatible Lustre on builders)...

You have control over what is installed on the build nodes by altering the `Dockerfile`s.  There's no reason you could not add the installation of Lustre to those, at least as an option, and then set that option in CI.  You can see in our existing `Dockerfile`s how we check optional flags and in `Jenkinsfile` how we have CI set them.

> Is it possible to generate RPMs in a local build ? And if yes, howto?

```bash
$ make -C utils/rpms rpms
```

But you of course have to have all of the build dependencies installed:

```bash
# yum-builddep utils/rpms/daos.spec
```

which means you have to have our repositories installed:

```bash
# cat <EOF > /etc/yum/repos.d/daos.repo
[repo.dc.hpdd.intel.com_repository_daos-stack-el-7-x86_64-stable-local]
name=created by dnf config-manager from https://repo.dc.hpdd.intel.com/repository/daos-stack-el-7-x86_64-stable-local
baseurl=https://repo.dc.hpdd.intel.com/repository/daos-stack-el-7-x86_64-stable-local
enabled=1
gpgcheck=False
EOF
```

> bitrot, hmm what do you mean??

https://en.wikipedia.org/wiki/Software_rot

> BTW, does the new style of my changes, more python/scons oriented, already looks better for you ??

Yes, as @jolivier23  said "much better!".

File: src/client/dfs/SConscript [hunk 13]
```diff
@@ -21,7 +21,7 @@ def scons():
             gotversion = True
     if gotversion is True:
         print("Building with Lustre bindings.")
-        denv.AppendUnique(CCFLAGS=['-DLUSTRE_INCLUDE'])
+        conf.env.AppendUnique(CCFLAGS=['-DLUSTRE_INCLUDE'])
```
Comment (jolivier23) [inline src/client/dfs/SConscript]: why this change?   denv looks right here

Comment (johannlombardi) [reply src/client/dfs/SConscript]: @bfaccini could you please address that?

File: src/client/dfs/SConscript [hunk 14]
```diff
@@ -7,49 +7,24 @@ def scons():
...
-    except ImportError:
-        pass
+            print("Lustre version is compatible")
+            gotversion = True
+    if gotversion is True:
+        print("Building with Lustre bindings.")
+        conf.env.AppendUnique(CCFLAGS=['-DLUSTRE_INCLUDE'])
+    else:
+        print("Not building with Lustre bindings.")
+    denv = conf.Finish()
```
Comment (jolivier23) [inline src/client/dfs/SConscript]: ah, I see.   I've never seen this before.

File: src/client/dfs/SConscript (hunk 13 above)
Comment (jolivier23) [reply src/client/dfs/SConscript]: I guess maybe it's fine but to be consistent with every other use of configure, it may be better to just use denv here and not set denv = conf.Finish()

Comment (bfaccini) [reply src/client/dfs/SConscript]: Well, my understanding is that conf.env is like clone of the current Environment when reaching Configure(), where it can be modified and then passed-back/returned using conf.Finish() ... BTW, it is always stated that Finish() has to be done, right ?

File: src/client/dfs/SConscript (hunk 14 above)
Comment (bfaccini) [reply src/client/dfs/SConscript]: I found this behaviour documented at "https://github.com/SCons/scons/wiki/SconsAutoconf"

File: src/client/dfs/SConscript (hunk 13 above)
Comment (jolivier23) [reply src/client/dfs/SConscript]: Yeah, we always do Finish.   I just don’t see a difference between the two approaches of modifying the Configure environment and modifying the external one directly.  They should both have exactly the same effect.   I’m fine either way really but just saying we do configure checks in a lot of other places and all call Finish without capturing the environment
